<img src="images/upGrad.png" alt="upGrad" align="Right" style="width: 200px;"/>
<img src="images/IIITB.jpeg" alt="IITB" align="Left" style="width: 200px;"/>

# Telecom Churn Prediction

<b>Author:</b> Karthik Premanand, Anish Mahapatra

<i>Telecom Churn Casestudy </i>

## Problem Statement

## Data Understanding and Exploration

In [1]:
# Importing the required packages

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import linear_model
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV

import os

# hide warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Removing the minimum display columns to 500
pd.set_option('display.max_columns', 500)
pd.set_option('display.max_rows', 500)

# Ignoring warnings
warnings.filterwarnings("ignore")

In [4]:
# Reading the dataset
telecom = pd.read_csv("Datasets/telecom_churn_data.csv")

In [5]:
# Sense Check of the Data
telecom.head()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   
3     7001204172        109             0.0             0.0             0.0   
4     7000142493        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0            6/30/2014            7/31/2014            8/31/2014   
1            6/30/2014            7/31/2014            8/31/2014   
2            6/30/2014            7/31/2014            8/31/2014   
3            6/30/2014            7/31/2014            8/31/2014   
4            6/30/2014            7/31/2014            8/31/2014   

  last_date_of_month_9   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
0            9/30/2014  197.385  214.816  213.803   21.100          NaN   
1            9/30/2014   34.047  355.074  268.321   86.285        24.11   
2            9/30/2014  167.690  189.058  210.226  290.714        11.54   
3            9/30/2014  221.338  251.102  508.054  389.500        99.91   
4            9/30/2014  261.636  309.876  238.174  163.426        50.31   

   onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
0          NaN         0.00          NaN           NaN           NaN   
1        78.68         7.68        18.34         15.74         99.84   
2        55.24        37.26        74.81        143.33        220.59   
3        54.39       310.98       241.71        123.31        109.01   
4       149.44        83.89        58.78         76.96         91.88   

   offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
0          0.00           NaN            NaN            NaN           0.00   
1        304.76         53.76            0.0           0.00           0.00   
2        208.36        118.91            0.0           0.00           0.00   
3         71.68        113.54            0.0          54.86          44.38   
4        124.26         45.81            0.0           0.00           0.00   

   roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
0            NaN            NaN            NaN           0.00            NaN   
1           0.00            0.0           0.00           0.00           0.00   
2          38.49            0.0           0.00           0.00          70.94   
3           0.00            0.0          28.09          39.04           0.00   
4           0.00            0.0           0.00           0.00           0.00   

   loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
0               NaN               NaN              0.00               NaN   
1             23.88             74.56              7.68             18.34   
2              7.19             28.74             13.58             14.39   
3             73.68             34.81             10.61             15.49   
4             50.31            149.44             83.89             58.78   

   loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
0               NaN               NaN              0.00               NaN   
1             11.51             75.94            291.86             53.76   
2             29.34             16.86             38.46             28.16   
3            107.43             83.21             22.46             65.46   
4             67.64             91.88            124.26             37.89   

   loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
0               NaN               NaN              0.00               NaN   
1              0.00              0.00              0.00              0.00   
2             24.11             21.79             15.61             22.24   
3              1.91              0.65              

In [6]:
# Viewing the shape of the data
telecom.shape

(99999, 226)

In [13]:
# Summary of the dataset
print(telecom.info(verbose=True))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 226 columns):
mobile_number               int64
circle_id                   int64
loc_og_t2o_mou              float64
std_og_t2o_mou              float64
loc_ic_t2o_mou              float64
last_date_of_month_6        object
last_date_of_month_7        object
last_date_of_month_8        object
last_date_of_month_9        object
arpu_6                      float64
arpu_7                      float64
arpu_8                      float64
arpu_9                      float64
onnet_mou_6                 float64
onnet_mou_7                 float64
onnet_mou_8                 float64
onnet_mou_9                 float64
offnet_mou_6                float64
offnet_mou_7                float64
offnet_mou_8                float64
offnet_mou_9                float64
roam_ic_mou_6               float64
roam_ic_mou_7               float64
roam_ic_mou_8               float64
roam_ic_mou_9               float6

In [16]:
# Printing all the columns with atleast one null value
telecom.columns[telecom.isna().any()].tolist()

['loc_og_t2o_mou',
 'std_og_t2o_mou',
 'loc_ic_t2o_mou',
 'last_date_of_month_7',
 'last_date_of_month_8',
 'last_date_of_month_9',
 'onnet_mou_6',
 'onnet_mou_7',
 'onnet_mou_8',
 'onnet_mou_9',
 'offnet_mou_6',
 'offnet_mou_7',
 'offnet_mou_8',
 'offnet_mou_9',
 'roam_ic_mou_6',
 'roam_ic_mou_7',
 'roam_ic_mou_8',
 'roam_ic_mou_9',
 'roam_og_mou_6',
 'roam_og_mou_7',
 'roam_og_mou_8',
 'roam_og_mou_9',
 'loc_og_t2t_mou_6',
 'loc_og_t2t_mou_7',
 'loc_og_t2t_mou_8',
 'loc_og_t2t_mou_9',
 'loc_og_t2m_mou_6',
 'loc_og_t2m_mou_7',
 'loc_og_t2m_mou_8',
 'loc_og_t2m_mou_9',
 'loc_og_t2f_mou_6',
 'loc_og_t2f_mou_7',
 'loc_og_t2f_mou_8',
 'loc_og_t2f_mou_9',
 'loc_og_t2c_mou_6',
 'loc_og_t2c_mou_7',
 'loc_og_t2c_mou_8',
 'loc_og_t2c_mou_9',
 'loc_og_mou_6',
 'loc_og_mou_7',
 'loc_og_mou_8',
 'loc_og_mou_9',
 'std_og_t2t_mou_6',
 'std_og_t2t_mou_7',
 'std_og_t2t_mou_8',
 'std_og_t2t_mou_9',
 'std_og_t2m_mou_6',
 'std_og_t2m_mou_7',
 'std_og_t2m_mou_8',
 'std_og_t2m_mou_9',
 'std_og_t2f_mou_6',

There are 99999 rows, 226 columns, in the dataset out of which 166 columns have null values

### Exploratory Data Analysis

#### Missing Value Analysis

In [70]:
df = telecom.copy(deep=True)

In [71]:
# Visulalizing the percentage of missing values
missing = (df.isnull().sum() / len(df)) * 100
missing_df = pd.DataFrame({'Feature': df.columns,
                                 'missing (%)': missing})

missing_df

Feature  missing (%)
mobile_number                        mobile_number     0.000000
circle_id                                circle_id     0.000000
loc_og_t2o_mou                      loc_og_t2o_mou     1.018010
std_og_t2o_mou                      std_og_t2o_mou     1.018010
loc_ic_t2o_mou                      loc_ic_t2o_mou     1.018010
last_date_of_month_6          last_date_of_month_6     0.000000
last_date_of_month_7          last_date_of_month_7     0.601006
last_date_of_month_8          last_date_of_month_8     1.100011
last_date_of_month_9          last_date_of_month_9     1.659017
arpu_6                                      arpu_6     0.000000
arpu_7                                      arpu_7     0.000000
arpu_8                                      arpu_8     0.000000
arpu_9                                      arpu_9     0.000000
onnet_mou_6                            onnet_mou_6     3.937039
onnet_mou_7                            onnet_mou_7     3.859039
onnet_mou_8                            onnet_mou_8     5.378054
onnet_mou_9                            onnet_mou_9     7.745077
offnet_mou_6                          offnet_mou_6     3.937039
offnet_mou_7                          offnet_mou_7     3.859039
offnet_mou_8                          offnet_mou_8     5.378054
offnet_mou_9                          offnet_mou_9     7.745077
roam_ic_mou_6                        roam_ic_mou_6     3.937039
roam_ic_mou_7                        roam_ic_mou_7     3.859039
roam_ic_mou_8                        roam_ic_mou_8     5.378054
roam_ic_mou_9                        roam_ic_mou_9     7.745077
roam_og_mou_6                        roam_og_mou_6     3.937039
roam_og_mou_7                        roam_og_mou_7     3.859039
roam_og_mou_8                        roam_og_mou_8     5.378054
roam_og_mou_9                        roam_og_mou_9     7.745077
loc_og_t2t_mou_6                  loc_og_t2t_mou_6     3.937039
loc_og_t2t_mou_7                  loc_og_t2t_mou_7     3.859039
loc_og_t2t_mou_8                  loc_og_t2t_mou_8     5.378054
loc_og_t2t_mou_9                  loc_og_t2t_mou_9     7.745077
loc_og_t2m_mou_6                  loc_og_t2m_mou_6     3.937039
loc_og_t2m_mou_7                  loc_og_t2m_mou_7     3.859039
loc_og_t2m_mou_8                  loc_og_t2m_mou_8     5.378054
loc_og_t2m_mou_9                  loc_og_t2m_mou_9     7.745077
loc_og_t2f_mou_6                  loc_og_t2f_mou_6     3.937039
loc_og_t2f_mou_7                  loc_og_t2f_mou_7     3.859039
loc_og_t2f_mou_8                  loc_og_t2f_mou_8     5.378054
loc_og_t2f_mou_9                  loc_og_t2f_mou_9     7.745077
loc_og_t2c_mou_6                  loc_og_t2c_mou_6     3.937039
loc_og_t2c_mou_7                  loc_og_t2c_mou_7     3.859039
loc_og_t2c_mou_8                  loc_og_t2c_mou_8     5.378054
loc_og_t2c_mou_9                  loc_og_t2c_mou_9     7.745077
loc_og_mou_6                          loc_og_mou_6     3.937039
loc_og_mou_7                          loc_og_mou_7     3.859039
loc_og_mou_8                          loc_og_mou_8     5.378054
loc_og_mou_9                          loc_og_mou_9     7.745077
std_og_t2t_mou_6                  std_og_t2t_mou_6     3.937039
std_og_t2t_mou_7                  std_og_t2t_mou_7     3.859039
std_og_t2t_mou_8                  std_og_t2t_mou_8     5.378054
std_og_t2t_mou_9                  std_og_t2t_mou_9     7.745077
std_og_t2m_mou_6                  std_og_t2m_mou_6     3.937039
std_og_t2m_mou_7                  std_og_t2m_mou_7     3.859039
std_og_t2m_mou_8                  std_og_t2m_mou_8     5.378054
std_og_t2m_mou_9                  std_og_t2m_mou_9     7.745077
std_og_t2f_mou_6                  std_og_t2f_mou_6     3.937039
std_og_t2f_mou_7                  std_og_t2f_mou_7     3.859039
std_og_t2f_mou_8                  std_og_t2f_mou_8     5.378054
std_og_t2f_mou_9                  std_og_t2f_mou_9     7.745077
std_og_t2c_mou_6                  std_og_t2c_mou_6     3.937039
std_og_t2c_

In [72]:
print(telecom['arpu_3g_6'].describe())

count    25153.000000
mean        89.555057
std        193.124653
min        -30.820000
25%          0.000000
50%          0.480000
75%        122.070000
max       6362.280000
Name: arpu_3g_6, dtype: float64


We realize that a lot of the columns have 'NA' values. There are two kinds of missing values here:

* Actual missing - 
    * date_of_last_rech_data_6
    * date_of_last_rech_data_7
    * date_of_last_rech_data_8
    * date_of_last_rech_data_9
    * total_rech_data_6	
    * total_rech_data_7	
    * total_rech_data_8	
    * total_rech_data_9
    * max_rech_data_6
    * max_rech_data_7
    * max_rech_data_8
    * max_rech_data_9
    * av_rech_amt_data_6
    * av_rech_amt_data_7
    * av_rech_amt_data_8
    * av_rech_amt_data_9
    
    
    
* Imputing the values with zero    
    * arpu_3g_6
    * arpu_3g_7
    * arpu_3g_8
    * arpu_3g_9
    * arpu_2g_6
    * arpu_2g_7
    * arpu_2g_8
    * arpu_2g_9
    
    
* Meaningful NA values - 
    * fb_user_6
    * fb_user_7
    * fb_user_8
    * fb_user_9
    * night_pck_user_6
    * night_pck_user_7
    * night_pck_user_8
    * night_pck_user_9
    * count_rech_2g_6
    * count_rech_2g_7
    * count_rech_2g_8
    * count_rech_2g_9
    * count_rech_3g_6
    * count_rech_3g_7
    * count_rech_3g_8
    * count_rech_3g_9
    

    


In [73]:
# Handling indicative missing values
df['arpu_3g_6'].fillna(0, inplace = True) 
df['arpu_3g_7'].fillna(0, inplace = True) 
df['arpu_3g_8'].fillna(0, inplace = True)
df['arpu_3g_9'].fillna(0, inplace = True) 
df['arpu_2g_6'].fillna(0, inplace = True) 
df['arpu_2g_7'].fillna(0, inplace = True)
df['arpu_2g_8'].fillna(0, inplace = True) 
df['arpu_2g_9'].fillna(0, inplace = True)
df['fb_user_6'].fillna(0, inplace = True) 
df['fb_user_7'].fillna(0, inplace = True) 
df['fb_user_8'].fillna(0, inplace = True)
df['fb_user_9'].fillna(0, inplace = True) 
df['night_pck_user_6'].fillna(0, inplace = True) 
df['night_pck_user_7'].fillna(0, inplace = True)
df['night_pck_user_8'].fillna(0, inplace = True) 
df['night_pck_user_9'].fillna(0, inplace = True)
df['count_rech_2g_6'].fillna(0, inplace = True) 
df['count_rech_2g_7'].fillna(0, inplace = True) 
df['count_rech_2g_8'].fillna(0, inplace = True)
df['count_rech_2g_9'].fillna(0, inplace = True) 
df['count_rech_3g_6'].fillna(0, inplace = True) 
df['count_rech_3g_7'].fillna(0, inplace = True)
df['count_rech_3g_8'].fillna(0, inplace = True) 
df['count_rech_3g_9'].fillna(0, inplace = True)

In [74]:
# Dropping columns with more than 70% Actual missing values

df = df.drop(['date_of_last_rech_data_6', 'date_of_last_rech_data_7','date_of_last_rech_data_8','date_of_last_rech_data_9','total_rech_data_6','total_rech_data_7','total_rech_data_8','total_rech_data_9','max_rech_data_6','max_rech_data_7','max_rech_data_8','max_rech_data_9','av_rech_amt_data_6','av_rech_amt_data_7','av_rech_amt_data_8','av_rech_amt_data_9'], axis=1)

In [75]:
# Visulalizing the percentage of missing values after handling the NA values
missing = (df.isnull().sum() / len(df)) * 100
missing_df = pd.DataFrame({'Feature': df.columns,
                                 'missing (%)': missing})

missing_df

Feature  missing (%)
mobile_number                mobile_number     0.000000
circle_id                        circle_id     0.000000
loc_og_t2o_mou              loc_og_t2o_mou     1.018010
std_og_t2o_mou              std_og_t2o_mou     1.018010
loc_ic_t2o_mou              loc_ic_t2o_mou     1.018010
last_date_of_month_6  last_date_of_month_6     0.000000
last_date_of_month_7  last_date_of_month_7     0.601006
last_date_of_month_8  last_date_of_month_8     1.100011
last_date_of_month_9  last_date_of_month_9     1.659017
arpu_6                              arpu_6     0.000000
arpu_7                              arpu_7     0.000000
arpu_8                              arpu_8     0.000000
arpu_9                              arpu_9     0.000000
onnet_mou_6                    onnet_mou_6     3.937039
onnet_mou_7                    onnet_mou_7     3.859039
onnet_mou_8                    onnet_mou_8     5.378054
onnet_mou_9                    onnet_mou_9     7.745077
offnet_mou_6                  offnet_mou_6     3.937039
offnet_mou_7                  offnet_mou_7     3.859039
offnet_mou_8                  offnet_mou_8     5.378054
offnet_mou_9                  offnet_mou_9     7.745077
roam_ic_mou_6                roam_ic_mou_6     3.937039
roam_ic_mou_7                roam_ic_mou_7     3.859039
roam_ic_mou_8                roam_ic_mou_8     5.378054
roam_ic_mou_9                roam_ic_mou_9     7.745077
roam_og_mou_6                roam_og_mou_6     3.937039
roam_og_mou_7                roam_og_mou_7     3.859039
roam_og_mou_8                roam_og_mou_8     5.378054
roam_og_mou_9                roam_og_mou_9     7.745077
loc_og_t2t_mou_6          loc_og_t2t_mou_6     3.937039
loc_og_t2t_mou_7          loc_og_t2t_mou_7     3.859039
loc_og_t2t_mou_8          loc_og_t2t_mou_8     5.378054
loc_og_t2t_mou_9          loc_og_t2t_mou_9     7.745077
loc_og_t2m_mou_6          loc_og_t2m_mou_6     3.937039
loc_og_t2m_mou_7          loc_og_t2m_mou_7     3.859039
loc_og_t2m_mou_8          loc_og_t2m_mou_8     5.378054
loc_og_t2m_mou_9          loc_og_t2m_mou_9     7.745077
loc_og_t2f_mou_6          loc_og_t2f_mou_6     3.937039
loc_og_t2f_mou_7          loc_og_t2f_mou_7     3.859039
loc_og_t2f_mou_8          loc_og_t2f_mou_8     5.378054
loc_og_t2f_mou_9          loc_og_t2f_mou_9     7.745077
loc_og_t2c_mou_6          loc_og_t2c_mou_6     3.937039
loc_og_t2c_mou_7          loc_og_t2c_mou_7     3.859039
loc_og_t2c_mou_8          loc_og_t2c_mou_8     5.378054
loc_og_t2c_mou_9          loc_og_t2c_mou_9     7.745077
loc_og_mou_6                  loc_og_mou_6     3.937039
loc_og_mou_7                  loc_og_mou_7     3.859039
loc_og_mou_8                  loc_og_mou_8     5.378054
loc_og_mou_9                  loc_og_mou_9     7.745077
std_og_t2t_mou_6          std_og_t2t_mou_6     3.937039
std_og_t2t_mou_7          std_og_t2t_mou_7     3.859039
std_og_t2t_mou_8          std_og_t2t_mou_8     5.378054
std_og_t2t_mou_9          std_og_t2t_mou_9     7.745077
std_og_t2m_mou_6          std_og_t2m_mou_6     3.937039
std_og_t2m_mou_7          std_og_t2m_mou_7     3.859039
std_og_t2m_mou_8          std_og_t2m_mou_8     5.378054
std_og_t2m_mou_9          std_og_t2m_mou_9     7.745077
std_og_t2f_mou_6          std_og_t2f_mou_6     3.937039
std_og_t2f_mou_7          std_og_t2f_mou_7     3.859039
std_og_t2f_mou_8          std_og_t2f_mou_8     5.378054
std_og_t2f_mou_9          std_og_t2f_mou_9     7.745077
std_og_t2c_mou_6          std_og_t2c_mou_6     3.937039
std_og_t2c_mou_7          std_og_t2c_mou_7     3.859039
std_og_t2c_mou_8          std_og_t2c_mou_8     5.378054
std_og_t2c_mou_9          std_og_t2c_mou_9     7.745077
std_og_mou_6                  std_og_mou_6     3.937039
std_og_mou_7                  std_og_mou_7     3.859039
std_og_mou_8                  std_og_mou_8     5.378054
std_og_mou_9                  std_og_mou_9     7.745077
isd_og_mou_6                  isd_og_mou_6     3.937039
isd_og_mou_7                  isd_og_mou_7     3.859039
isd

No more columns will be removed as all the remaining missing values are in the acceptable range

#### Calculating derived metrics before dropping date objects

In [76]:
### Calculating the number of days since last recharge for a particular month
df['last_date_of_month_6'] = pd.to_datetime(df['last_date_of_month_6'])
df['date_of_last_rech_6'] = pd.to_datetime(df['date_of_last_rech_6'])
df['day_of_last_rech_6'] = (df['last_date_of_month_6'] - df['date_of_last_rech_6']) / np.timedelta64(1, 'D')
df['week_of_last_rech_6'] = df['day_of_last_rech_6'] % 7

df['last_date_of_month_7'] = pd.to_datetime(df['last_date_of_month_7'])
df['date_of_last_rech_7'] = pd.to_datetime(df['date_of_last_rech_7'])
df['day_of_last_rech_7'] = (df['last_date_of_month_7'] - df['date_of_last_rech_7']) / np.timedelta64(1, 'D')
df['week_of_last_rech_7'] = df['day_of_last_rech_7'] % 7

df['last_date_of_month_8'] = pd.to_datetime(df['last_date_of_month_8'])
df['date_of_last_rech_8'] = pd.to_datetime(df['date_of_last_rech_8'])
df['day_of_last_rech_8'] = (df['last_date_of_month_8'] - df['date_of_last_rech_8']) / np.timedelta64(1, 'D')
df['week_of_last_rech_8'] = df['day_of_last_rech_8'] % 7

df['last_date_of_month_9'] = pd.to_datetime(df['last_date_of_month_9'])
df['date_of_last_rech_9'] = pd.to_datetime(df['date_of_last_rech_9'])
df['day_of_last_rech_9'] = (df['last_date_of_month_9'] - df['date_of_last_rech_9']) / np.timedelta64(1, 'D')
df['week_of_last_rech_9'] = df['day_of_last_rech_9'] % 7

In [78]:
### Dropping the date objects

df = df.drop(['last_date_of_month_6','last_date_of_month_7','last_date_of_month_8','last_date_of_month_9','date_of_last_rech_6','date_of_last_rech_7','date_of_last_rech_8','date_of_last_rech_9'], axis=1)


KeyError: "['last_date_of_month_6' 'last_date_of_month_7' 'last_date_of_month_8'\n 'last_date_of_month_9' 'date_of_last_rech_6' 'date_of_last_rech_7'\n 'date_of_last_rech_8' 'date_of_last_rech_9'] not found in axis"

In [79]:
### Dropping the mobile number

df = df.drop(['mobile_number'],axis= 1)

In [80]:
# Visulalizing the percentage of missing values after handling the date objects
missing = (df.isnull().sum() / len(df)) * 100
missing_df = pd.DataFrame({'Feature': df.columns,
                                 'missing (%)': missing})

missing_df

Feature  missing (%)
circle_id                      circle_id     0.000000
loc_og_t2o_mou            loc_og_t2o_mou     1.018010
std_og_t2o_mou            std_og_t2o_mou     1.018010
loc_ic_t2o_mou            loc_ic_t2o_mou     1.018010
arpu_6                            arpu_6     0.000000
arpu_7                            arpu_7     0.000000
arpu_8                            arpu_8     0.000000
arpu_9                            arpu_9     0.000000
onnet_mou_6                  onnet_mou_6     3.937039
onnet_mou_7                  onnet_mou_7     3.859039
onnet_mou_8                  onnet_mou_8     5.378054
onnet_mou_9                  onnet_mou_9     7.745077
offnet_mou_6                offnet_mou_6     3.937039
offnet_mou_7                offnet_mou_7     3.859039
offnet_mou_8                offnet_mou_8     5.378054
offnet_mou_9                offnet_mou_9     7.745077
roam_ic_mou_6              roam_ic_mou_6     3.937039
roam_ic_mou_7              roam_ic_mou_7     3.859039
roam_ic_mou_8              roam_ic_mou_8     5.378054
roam_ic_mou_9              roam_ic_mou_9     7.745077
roam_og_mou_6              roam_og_mou_6     3.937039
roam_og_mou_7              roam_og_mou_7     3.859039
roam_og_mou_8              roam_og_mou_8     5.378054
roam_og_mou_9              roam_og_mou_9     7.745077
loc_og_t2t_mou_6        loc_og_t2t_mou_6     3.937039
loc_og_t2t_mou_7        loc_og_t2t_mou_7     3.859039
loc_og_t2t_mou_8        loc_og_t2t_mou_8     5.378054
loc_og_t2t_mou_9        loc_og_t2t_mou_9     7.745077
loc_og_t2m_mou_6        loc_og_t2m_mou_6     3.937039
loc_og_t2m_mou_7        loc_og_t2m_mou_7     3.859039
loc_og_t2m_mou_8        loc_og_t2m_mou_8     5.378054
loc_og_t2m_mou_9        loc_og_t2m_mou_9     7.745077
loc_og_t2f_mou_6        loc_og_t2f_mou_6     3.937039
loc_og_t2f_mou_7        loc_og_t2f_mou_7     3.859039
loc_og_t2f_mou_8        loc_og_t2f_mou_8     5.378054
loc_og_t2f_mou_9        loc_og_t2f_mou_9     7.745077
loc_og_t2c_mou_6        loc_og_t2c_mou_6     3.937039
loc_og_t2c_mou_7        loc_og_t2c_mou_7     3.859039
loc_og_t2c_mou_8        loc_og_t2c_mou_8     5.378054
loc_og_t2c_mou_9        loc_og_t2c_mou_9     7.745077
loc_og_mou_6                loc_og_mou_6     3.937039
loc_og_mou_7                loc_og_mou_7     3.859039
loc_og_mou_8                loc_og_mou_8     5.378054
loc_og_mou_9                loc_og_mou_9     7.745077
std_og_t2t_mou_6        std_og_t2t_mou_6     3.937039
std_og_t2t_mou_7        std_og_t2t_mou_7     3.859039
std_og_t2t_mou_8        std_og_t2t_mou_8     5.378054
std_og_t2t_mou_9        std_og_t2t_mou_9     7.745077
std_og_t2m_mou_6        std_og_t2m_mou_6     3.937039
std_og_t2m_mou_7        std_og_t2m_mou_7     3.859039
std_og_t2m_mou_8        std_og_t2m_mou_8     5.378054
std_og_t2m_mou_9        std_og_t2m_mou_9     7.745077
std_og_t2f_mou_6        std_og_t2f_mou_6     3.937039
std_og_t2f_mou_7        std_og_t2f_mou_7     3.859039
std_og_t2f_mou_8        std_og_t2f_mou_8     5.378054
std_og_t2f_mou_9        std_og_t2f_mou_9     7.745077
std_og_t2c_mou_6        std_og_t2c_mou_6     3.937039
std_og_t2c_mou_7        std_og_t2c_mou_7     3.859039
std_og_t2c_mou_8        std_og_t2c_mou_8     5.378054
std_og_t2c_mou_9        std_og_t2c_mou_9     7.745077
std_og_mou_6                std_og_mou_6     3.937039
std_og_mou_7                std_og_mou_7     3.859039
std_og_mou_8                std_og_mou_8     5.378054
std_og_mou_9                std_og_mou_9     7.745077
isd_og_mou_6                isd_og_mou_6     3.937039
isd_og_mou_7                isd_og_mou_7     3.859039
isd_og_mou_8                isd_og_mou_8     5.378054
isd_og_mou_9                isd_og_mou_9     7.745077
spl_og_mou_6                spl_og_mou_6     3.937039
spl_og_mou_7                spl_og_mou_7     3.859039
spl_og_mou_8                spl_og_mou_8     5.378054
spl_og_mou_9                spl_og_mou_9     7.745077
og_others_6                  og_others_6     3.937039
og_others_7                  og_other

#### Iterative imputation for missing values between 0-8 %

Using sklearn's iterative imputer to calculate missing values for columns in range 1-8%.
<br>**Note:** The initialization strategy depends on the respective column. The estimator used is Bayesian Ridge.
The columns and initialization method are as follows:
- MOU : mean
- OTHERS, DAY_OF_LAST_RECH, WEEK_OF_LAST_RECH : median

In [1]:
# Using sklearn's iterative imputer to calculate missing values for columns in range 1-8%.
## Note: The initialization strategy depends on the respective column. The estimator used is Bayesian Ridge.

from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

imp_mean = IterativeImputer(missing_values=np.nan, sample_posterior=False, 
                                 max_iter=10, tol=0.001, 
                                 n_nearest_features=4, initial_strategy='mean')
imp_mean.fit(df)

NameError: name 'np' is not defined

In [ ]:
df = pd.DataFrame(data=imp_mean.transform(df),
                  columns = list(df.columns.values),
                             dtype='int')

In [83]:
# Visulalizing the percentage of missing values after handling the date objects
missing = (df.isnull().sum() / len(df)) * 100
missing_df = pd.DataFrame({'Feature': df.columns,
                                 'missing (%)': missing})

missing_df

Feature  missing (%)
circle_id                      circle_id          0.0
loc_og_t2o_mou            loc_og_t2o_mou          0.0
std_og_t2o_mou            std_og_t2o_mou          0.0
loc_ic_t2o_mou            loc_ic_t2o_mou          0.0
arpu_6                            arpu_6          0.0
arpu_7                            arpu_7          0.0
arpu_8                            arpu_8          0.0
arpu_9                            arpu_9          0.0
onnet_mou_6                  onnet_mou_6          0.0
onnet_mou_7                  onnet_mou_7          0.0
onnet_mou_8                  onnet_mou_8          0.0
onnet_mou_9                  onnet_mou_9          0.0
offnet_mou_6                offnet_mou_6          0.0
offnet_mou_7                offnet_mou_7          0.0
offnet_mou_8                offnet_mou_8          0.0
offnet_mou_9                offnet_mou_9          0.0
roam_ic_mou_6              roam_ic_mou_6          0.0
roam_ic_mou_7              roam_ic_mou_7          0.0
roam_ic_mou_8              roam_ic_mou_8          0.0
roam_ic_mou_9              roam_ic_mou_9          0.0
roam_og_mou_6              roam_og_mou_6          0.0
roam_og_mou_7              roam_og_mou_7          0.0
roam_og_mou_8              roam_og_mou_8          0.0
roam_og_mou_9              roam_og_mou_9          0.0
loc_og_t2t_mou_6        loc_og_t2t_mou_6          0.0
loc_og_t2t_mou_7        loc_og_t2t_mou_7          0.0
loc_og_t2t_mou_8        loc_og_t2t_mou_8          0.0
loc_og_t2t_mou_9        loc_og_t2t_mou_9          0.0
loc_og_t2m_mou_6        loc_og_t2m_mou_6          0.0
loc_og_t2m_mou_7        loc_og_t2m_mou_7          0.0
loc_og_t2m_mou_8        loc_og_t2m_mou_8          0.0
loc_og_t2m_mou_9        loc_og_t2m_mou_9          0.0
loc_og_t2f_mou_6        loc_og_t2f_mou_6          0.0
loc_og_t2f_mou_7        loc_og_t2f_mou_7          0.0
loc_og_t2f_mou_8        loc_og_t2f_mou_8          0.0
loc_og_t2f_mou_9        loc_og_t2f_mou_9          0.0
loc_og_t2c_mou_6        loc_og_t2c_mou_6          0.0
loc_og_t2c_mou_7        loc_og_t2c_mou_7          0.0
loc_og_t2c_mou_8        loc_og_t2c_mou_8          0.0
loc_og_t2c_mou_9        loc_og_t2c_mou_9          0.0
loc_og_mou_6                loc_og_mou_6          0.0
loc_og_mou_7                loc_og_mou_7          0.0
loc_og_mou_8                loc_og_mou_8          0.0
loc_og_mou_9                loc_og_mou_9          0.0
std_og_t2t_mou_6        std_og_t2t_mou_6          0.0
std_og_t2t_mou_7        std_og_t2t_mou_7          0.0
std_og_t2t_mou_8        std_og_t2t_mou_8          0.0
std_og_t2t_mou_9        std_og_t2t_mou_9          0.0
std_og_t2m_mou_6        std_og_t2m_mou_6          0.0
std_og_t2m_mou_7        std_og_t2m_mou_7          0.0
std_og_t2m_mou_8        std_og_t2m_mou_8          0.0
std_og_t2m_mou_9        std_og_t2m_mou_9          0.0
std_og_t2f_mou_6        std_og_t2f_mou_6          0.0
std_og_t2f_mou_7        std_og_t2f_mou_7          0.0
std_og_t2f_mou_8        std_og_t2f_mou_8          0.0
std_og_t2f_mou_9        std_og_t2f_mou_9          0.0
std_og_t2c_mou_6        std_og_t2c_mou_6          0.0
std_og_t2c_mou_7        std_og_t2c_mou_7          0.0
std_og_t2c_mou_8        std_og_t2c_mou_8          0.0
std_og_t2c_mou_9        std_og_t2c_mou_9          0.0
std_og_mou_6                std_og_mou_6          0.0
std_og_mou_7                std_og_mou_7          0.0
std_og_mou_8                std_og_mou_8          0.0
std_og_mou_9                std_og_mou_9          0.0
isd_og_mou_6                isd_og_mou_6          0.0
isd_og_mou_7                isd_og_mou_7          0.0
isd_og_mou_8                isd_og_mou_8          0.0
isd_og_mou_9                isd_og_mou_9          0.0
spl_og_mou_6                spl_og_mou_6          0.0
spl_og_mou_7                spl_og_mou_7          0.0
spl_og_mou_8                spl_og_mou_8          0.0
spl_og_mou_9                spl_og_mou_9          0.0
og_others_6                  og_others_6          0.0
og_others_7                  og_other

No more missing values, no categorical data to handle

### Filter high value customers

80-20 rule - 80% revenue generated by 20% customers


months 6,7 - good phase

Filteration is based on usage and not revenue generated with equal weightage given to both calling minutes and data usage (calling in terms of mins, data usage in terms of MB)
 
total_calling_usage_6 = total_og_mou_6 + total_ic_mou_6
<br>
total_data_usage_6 = vol_2g_mb_6 + vol_3g_mb_6
<br>
total_calling_usage_7 = total_og_mou_7 + total_ic_mou_7
<br>
total_data_usage_7 = vol_2g_mb_7 + vol_3g_mb_7

<br>
Total usage for months 6 and 7:
<br>
total_usage_6 = total_calling_usage_6 + total_data_usage_6
<br>
total_usage_7 = total_calling_usage_7 + total_data_usage_7

Average usage for months 6 and 7:
<br>
avg_usage_6_7 = (total_usage_6 + total_usage_7)/2

In [89]:
## Generating avg usage for customers for the 'good' months 6 & 7

df['total_calling_usage_6'] = df['total_og_mou_6'] + df['total_ic_mou_6']
df['total_data_usage_6'] = df['vol_2g_mb_6'] + df['vol_3g_mb_6']

df['total_calling_usage_7'] = df['total_og_mou_7'] + df['total_ic_mou_7']
df['total_data_usage_7'] = df['vol_2g_mb_7'] + df['vol_3g_mb_7']

df['total_usage_6'] = df['total_calling_usage_6'] + df['total_data_usage_6']
df['total_usage_7'] = df['total_calling_usage_7'] + df['total_data_usage_7']

df['avg_usage_6_7'] = (df['total_usage_6'] + df['total_usage_7']) / 2


In [92]:
df['avg_usage_6_7'].quantile(0.7)

762.0

In [93]:
df['hvc_flag'] = [1 if x >= df['avg_usage_6_7'].quantile(0.7) else 0 for x in df['avg_usage_6_7']]

### Deriving target churn variable

Calculate churn variable based on month 9

